lib/elastic_apm/central_config.rb

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+require 'elastic_apm/central_config/cache_control'
+
+module ElasticAPM
+  # @api private
+  class CentralConfig
+    include Logging
+
+    # @api private
+    class ResponseError < InternalError
+      def initialize(response)
+        @response = response
+      end
+
+      attr_reader :response
+    end
+    class NotFoundError < ResponseError; end
+    class ServerError < ResponseError; end
+
+    DEFAULT_MAX_AGE = 300
+
+    def initialize(config, default_polling_interval: 300)
+      @config = config
+      @service_info = {
+        'service.name': config.service_name,
+        'service.environment': config.environment
+      }.to_json
+      @default_polling_interval = default_polling_interval
+    end
+
+    attr_reader :config, :task
+
+    def start
+      fetch_and_apply_config
+    end
+
+    def fetch_and_apply_config
+      @task&.cancel
+      @task =
+        Concurrent::Promise
+        .execute(&method(:fetch_config))
+        .on_success(&method(:handle_success))
+        .rescue(&method(:handle_error))
+    end
+
+    def stop
+      @task&.cancel
+    end
+
+    def fetch_config
+      resp = perform_request
+
+      case resp.status
+      when 200..299
+        resp
+      when 404
+        raise NotFoundError, resp
+      when 500
+        raise ServerError, resp
+      end
+    end
+
+    private
+
+    def handle_success(resp)
+      update = JSON.parse(resp.body)
+      config.assign(update)
+
+      info 'Updated config from APM Server'
+
+      schedule_next_fetch(resp)
+
+      true
+    end
+
+    def handle_error(error)
+      error(
+        'Failed fetching config: %s, trying again in %d seconds',
+        error.response.body, DEFAULT_MAX_AGE
```

Comment:
`DEFAULT_MAX_AGE` is not correct here, as usually you will follow the `max-age` from the response cache-control header.